### ISSUE TITLE
deadlock found

### ISSUE THREAD

**Issue by DCjanus** [mattsse/chromiumoxide#50]
Hi, I was trying to impl a HTML-to-PNG service on this crate, and with your great work, the service is almost done, but unfortunately, before service to be used in production environment, we would test it with million random requests. At first, everything looks fine, but after some time, maybe an hour, maybe a day, service would hang forever.

Each sevice instance create 8 chromium instances, each request to my service would generate HTML to be render, with `debug=true` in query string, generated HTML would be respond, without this flag, my service would random choose an chromium instance and set content for it.

When this issue happend, my service thread is busy(100% cpu used in `htop`), and some( 1 ~ 2 ) chromium threads with 30% ~ 70% cpu used, remain of them are sleeping. Requests with `debug=true`, which means it respond without calling `chromiumoxide`, response successfully. Requests without `debug=true`, hang forever without any error, even if I've set [request_timeout](https://docs.rs/chromiumoxide/0.2.0/chromiumoxide/browser/struct.BrowserConfigBuilder.html#method.request_timeout) to 5 secs.

GDB attach my service thread, paused it and print thread backtrace, paused it and print thread backtrace ... . Repeat this a dozen times, here is my result:

```
(gdb) c
Continuing.

Thread 1 "fireshot" received signal SIGINT, Interrupt.
0x0000564790dd0032 in chromiumoxide::cmd::CommandChain::poll ()
(gdb) c
Continuing.

Thread 1 "fireshot" received signal SIGINT, Interrupt.
0x0000564790e8d12a in chromiumoxide::handler::target::Target::poll ()
(gdb) c
Continuing.

Thread 1 "fireshot" received signal SIGINT, Interrupt.
0x0000564790dd0032 in chromiumoxide::cmd::CommandChain::poll ()
(gdb) c
Continuing.

Thread 1 "fireshot" received signal SIGINT, Interrupt.
0x0000564790dcff9c in chromiumoxide::cmd::CommandChain::poll ()
(gdb) c
Continuing.

Thread 1 "fireshot" received signal SIGINT, Interrupt.
0x0000564790dd0214 in chromiumoxide::cmd::CommandChain::poll ()
(gdb) c
Continuing.

Thread 1 "fireshot" received signal SIGINT, Interrupt.
0x0000564790e8d17f in chromiumoxide::handler::target::Target::poll ()
(gdb) c
Continuing.

Thread 1 "fireshot" received signal SIGINT, Interrupt.
0x0000564790e8d192 in chromiumoxide::handler::target::Target::poll ()
(gdb) c
Continuing.

Thread 1 "fireshot" received signal SIGINT, Interrupt.
0x0000564790dd0032 in chromiumoxide::cmd::CommandChain::poll ()
(gdb) c
Continuing.

Thread 1 "fireshot" received signal SIGINT, Interrupt.
0x0000564790dd0032 in chromiumoxide::cmd::CommandChain::poll ()
(gdb) c
Continuing.

Thread 1 "fireshot" received signal SIGINT, Interrupt.
0x0000564790dcff94 in chromiumoxide::cmd::CommandChain::poll ()
(gdb) c
Continuing.

Thread 1 "fireshot" received signal SIGINT, Interrupt.
0x0000564790e8d1a1 in chromiumoxide::handler::target::Target::poll ()
(gdb) c
Continuing.

Thread 1 "fireshot" received signal SIGINT, Interrupt.
0x0000564790e8d101 in chromiumoxide::handler::target::Target::poll ()
(gdb) c
Continuing.

Thread 1 "fireshot" received signal SIGINT, Interrupt.
0x00005647913a220e in core::tuple::<impl core::cmp::Ord for (A,B)>::cmp () at /rustc/88f19c6dab716c6281af7602e30f413e809c5974/library/core/src/tuple.rs:58
58	/rustc/88f19c6dab716c6281af7602e30f413e809c5974/library/core/src/tuple.rs: No such file or directory.
(gdb) c
Continuing.

Thread 1 "fireshot" received signal SIGINT, Interrupt.
0x0000564790e8d187 in chromiumoxide::handler::target::Target::poll ()
(gdb) c
Continuing.

Thread 1 "fireshot" received signal SIGINT, Interrupt.
0x0000564790e8d18d in chromiumoxide::handler::target::Target::poll ()
(gdb) c
Continuing.

Thread 1 "fireshot" received signal SIGINT, Interrupt.
0x0000564790dd0214 in chromiumoxide::cmd::CommandChain::poll ()
(gdb) c
Continuing.

Thread 1 "fireshot" received signal SIGINT, Interrupt.
0x0000564790dd0032 in chromiumoxide::cmd::CommandChain::poll ()
```

**Comment by DCjanus** [comment]
I'm trying to fix this issue, but I am not familiar enough with this part of the code for now, if you can help me with this issue, that would be the best.

**Comment by mattsse** [comment]
> Each sevice instance create 8 chromium instances, each request to my service would generate HTML to be render, with debug=true in query string, generated HTML would be respond, without this flag, my service would random choose an chromium instance and set content for it.

Sorry I'm afraid I did not understand that part. So I have a couple of questions myself:
* you're launching 8 different instances of `chromium`? Each with a new remote-debug-port? 
* you're connecting `chromiumoxide::Browser` to each of them separately? 
* with `debug=true` you mean the rust [profile](https://doc.rust-lang.org/cargo/reference/profiles.html) right?

If so, then I might know what happens here. There is a difference in overflow handling between rust's release and debug mode: http://huonw.github.io/blog/2016/04/myths-and-legends-about-integer-overflow-in-rust/
basically in release mode there aren't any overflows.
So there seems to a code path that loops forever, from gdb output, this is the suspect: https://github.com/mattsse/chromiumoxide/blob/main/src/handler/target.rs#L273-L490

Maybe this here: https://github.com/mattsse/chromiumoxide/blob/main/src/cmd.rs#L145 ?

So my hunch is, when you run a lot of commands that may fail, perhaps they don't time out or aren't cleared, so maybe there is some weird edge case that causes so much back pressure to the `CommandList` that `target::poll` loops for ever? 
Just a hunch.
Those sections could definitely use some `log` statements. I'd start there.

**Comment by DCjanus** [comment]
> you're launching 8 different instances of chromium? Each with a new remote-debug-port?

Yes, I launch `chromium` via `--remote-debugging-port=0`

> you're connecting chromiumoxide::Browser to each of them separately?

Yes, and `actix_web::spawn` a task to consume everything from `Handler`.

> with debug=true you mean the rust profile right?

No, it's some logic in my service, I was just trying to explain, it's not hang in my code, but, my poor English make you confused.

When I found service is hang, I stopped requests to this instance, and wait for 10 hours, still hang there, and the output of GDB is similar. So, I do belive that not about back pressure.

Here is my `htop` output, I hope this would be helpful.

![图片](https://user-images.githubusercontent.com/15802727/119431637-8368ab00-bd45-11eb-9c79-caed30e1cbc6.png)

**Comment by DCjanus** [comment]
I've killed all `chromiumoxide` process, and nothing changed, hang without any error.

![图片](https://user-images.githubusercontent.com/15802727/119431784-c460bf80-bd45-11eb-8ead-3bb10bd239e9.png)

**Comment by mattsse** [comment]
I fixed a bug that was related to `Commandchain::poll` in #56 that resulted in timed out requests weren't cleared and `Commandchain::poll` called continuously. 
Maybe this fixes this issue as well

**Comment by DCjanus** [comment]
Good to know, I would like to ask my QA partner test again.

**Comment by DCjanus** [comment]
Hi, after so many days, finally I found the root cause.

In my case, my service would create and drop Browser again and agin, and without `waitpid`, there would be a lot of zombie process.

Those zombie process took all available pid, and we could never create new Browser, service looks like hang.

Sorry for my noise, and thanks for your great work.

**Comment by DCjanus** [comment]
This how I fix my problem:

```Rust
  #[cfg(target_family = "unix")]
  unsafe {
      nix::sys::signal::signal(
          nix::sys::signal::Signal::SIGCHLD,
          nix::sys::signal::SigHandler::SigIgn,
      )
      .expect("set SIGCHLD to SIG_IGN");
  }
```
